### DISCUSSION TITLE
Can the token extension attribute also support a list beyond str and bool values?

### DISCUSSION THREAD

**Discussion by lingvisa** [explosion/spaCy#8762] [Help: Coding & Implementations] [answered]
In this example:

https://spacy.io/usage/processing-pipelines

A doc extension can support a list as value, i.e.:

```
# Register custom extension on the Doc
        if not Doc.has_extension("acronyms"):
            Doc.set_extension("acronyms", default=[])
```

Then add values:
`doc._.acronyms.append((span, acronym))`

This doesn't seem to work for token extension. The following will report error:
```
Token.set_extension('featureList', default=[])
...
token._.featureList.append(tag)
```
I hope to be able to use a list as extension attribute's value, beyond str and bool:

`[{"LOWER": "apple"}, {"_": {"featureList": {"IN": "Company", "Fruit"}}}]`

**Comment by polm** [comment, marked as answer]
I was able to create a token extension with a list default without issue. 

```
import spacy
from spacy.tokens import Token

Token.set_extension("stuff", default=[])

nlp = spacy.blank("en")

doc = nlp("I like text.")
doc[0]._.stuff.append("blah")
doc[1]._.stuff.append("foo")

for tok in doc:
    print(tok, tok._.stuff, sep="\t")
```

What version of spaCy are you using and what error did you get?

That said I am not sure this will work with the Matcher, though I would need to look at that.

**Reply by lingvisa** [reply]
Thanks for the example, and I will try again.

**Comment by lingvisa** [comment]
The extension attribute as a list works for token as well, but the Matcher doesn't seem to work with extension attributes as list, as shown below:

```
import spacy
from spacy.matcher import Matcher
from spacy.tokens import Token

nlp = spacy.load("en_core_web_sm")
matcher = Matcher(nlp.vocab)

# Add pattern for valid hashtag, i.e. '#' plus any ASCII token
matcher.add("HASHTAG", [[{"ORTH": "#"}, {"IS_ASCII": True}]])

# Register token extension
Token.set_extension("is_hashtag", default=[])

doc = nlp("Hello world 😀 #MondayMotivation")
matches = matcher(doc)
hashtags = []
for match_id, start, end in matches:
    if doc.vocab.strings[match_id] == "HASHTAG":
        hashtags.append(doc[start:end])
with doc.retokenize() as retokenizer:
    for span in hashtags:
        retokenizer.merge(span)
        for token in span:
            token._.is_hashtag.append("true")

for token in doc:
    print(token.text, token._.is_hashtag)

## Now I want to match based on 'is_hashtag' as a list
matcher.add('HASH_LIST', [[{"_": {"is_hashtag": {"IN": ["true"]}}}]])
matches = matcher(doc)
for match in matches:
    print(match)
```

This gives this error:

```
Traceback (most recent call last):
  File "/Users/congminmin/nlp/wb-nlp-tool/nlu-weibo/wbnlu/tests/random_test.py", line 30, in <module>
    matches = matcher(doc)
  File "spacy/matcher/matcher.pyx", line 249, in spacy.matcher.matcher.Matcher.__call__
  File "spacy/matcher/matcher.pyx", line 378, in spacy.matcher.matcher.find_matches
  File "spacy/matcher/matcher.pyx", line 415, in spacy.matcher.matcher.transition_states
  File "spacy/matcher/matcher.pyx", line 526, in spacy.matcher.matcher.update_predicate_cache
  File "spacy/matcher/matcher.pyx", line 880, in spacy.matcher.matcher._SetPredicate.__call__
TypeError: unhashable type: 'list'
```

**Reply by polm** [reply]
Sorry, it looks like the Matcher operators don't support lists - I don't think we'd considered that usage pattern. I looked at the `ISSUBSET` and `ISSUPERSET` operators and I don't think they're right for this either. For now I think the best you can do is post-process results to check your list attributes. 

I'm not sure how complicated it would be to add, but I could definitely see us adding a `CONTAINS` operator for checking list properties. I'll discuss it with the team and see if there are any potential downsides or reasons we can't support it.

**Reply by adrianeboyd** [reply]
I had a look and the changes needed to support list values and add `IS_INTERSECT` are pretty minimal: #8784

**Comment by lingvisa** [comment]
If this can be supported in some way, that would be great.

**Reply by polm** [reply]
As Adriane mentioned, this has been added in #8784 and will likely be in our next release. 

Thanks for bringing it to our attention!

**Reply by lingvisa** [reply]
So what's the syntax of this in Matcher. Is it like this below, for example?

`{"_": {"is_hashtag": {"IN": ["true", "false"]}}}`

**Reply by polm** [reply]
The keyword it uses is `INTERSECTS`, [the tests][tests] might be a good reference. For example:

```
    # INTERSECTS with a list value
    Token.set_extension("ext", default=[])
    matcher = Matcher(en_vocab)
    pattern = [{"_": {"ext": {"INTERSECTS": ["A", "C"]}}}]
    matcher.add("M", [pattern])
    doc = Doc(en_vocab, words=["a", "b", "c"])
    doc[0]._.ext = ["A", "B"]
    assert len(matcher(doc)) == 1
```

Looking at your example code, I do not understand why `is_hashtag` is a list and not just a boolean. 

[tests]: https://github.com/explosion/spaCy/blob/56d4d87aebdaf3383c9ae8f482f47edd423de1cd/spacy/tests/matcher/test_matcher_api.py#L331

**Reply by lingvisa** [reply]
Understood. The is_hashtag example was copied from spacy doc and I slightly modified to illustrate the need, though not good example.
